Gems/Atom/Feature/Common/Code/Source/Mesh/ModelReloaderSystem.cpp

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Source/Mesh/ModelReloaderSystem.h>
+#include <Source/Mesh/ModelReloader.h>
+#include <AzCore/std/parallel/scoped_lock.h>
+
+namespace AZ
+{
+    namespace Render
+    {
+        void ModelReloaderSystem::ReloadModel(Data::Asset<RPI::ModelAsset> modelAsset, ModelReloadedEvent::Handler& onReloadedEventHandler)
+        {
+            AZStd::scoped_lock lock(m_pendingReloadMutex);
+            if (m_pendingReloads.count(modelAsset.GetId()) == 0)
+            {
+                ModelReloader* reloader = new ModelReloader(modelAsset, m_removeModelHandler);
```

Comment:
The ModelReloader deletes itself after calling RemoveReloader via the event specifically to work around this. Calling delete in RemoveReloader would delete the object while it is still in the middle of calling the event, which seems like a bad idea. Deleting itself allows the delete to occur after the RemoveReloader event is finished.

The other approach I thought of was to make the callback queue RemoveReloader, so that the ModelReloader finishes the event callback, and then at some later point in the future, like the next system tick, the ModelReloader could be removed + deleted by the ReloaderSystem. I didn't go that route since it creates a window during which any new requests to reload the model would find that there is already a pending reload, even though the pending reload is actually an old reload that is about to be removed. But I don't think that's a real problem in practice, since I don't believe it would ever occur that the asset is changed on two consecutive frames, since it takes time to process the asset. So the only real reason I didn't go that route was because I already had a working solution with the other approach.

Reply:
How about putting a comment in ModelReloaderSystem::RemoveReloader to indicate where to look for the memory cleanup? That will reduce the "surprise factor".
I would also ask @kberg-amzn whether it's allowed to delete the caller from within the callback.

Reply:
Added a comment as you suggested.